### PR TITLE
Crossplane RBAC Manager design doc

### PR DESCRIPTION
**Pull request by negz**
### Description of your changes

This PR introduces a design document for an "RBAC manager" process that runs alongside Crossplane, creating and managing (in some cases opinionated) RBAC roles for providers, composite resources, and composite resource claims. This RBAC manager process would replace similar functionality available in the contemporary package manager.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [ ] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [inline design/design-doc-rbac-manager.md:140]: We will also need the following permissions for the existing package manager, as well as the new package manager (which we may be able to be paired down after removal of the existing):
```yaml
rules:
  - apiGroups: ["extensions", "apps"]
    resources: ["deployments"]
    verbs: ["get", "list", "create", "update", "patch", "delete", "deletecollection"]
  - apiGroups: ["batch", "extensions"]
    resources: ["jobs"]
    verbs: ["get", "list", "create", "update", "patch", "delete", "deletecollection"]
  - apiGroups: [""]
    resources: ["secrets"]
    verbs: ["get", "list", "create", "update", "patch", "delete"]
  - apiGroups: [""]
    resources: ["pods"]
    verbs: ["get", "list"]
  - apiGroups: [""]
    resources: ["pods/log"]
    verbs: ["get"]
```
However, these are only needed for a `Role` and can be bound with a `RoleBinding` as the package manager only operates in the namespace Crossplane is running in. Note that these are the [same privileges the package manager has in hosted mode](https://github.com/crossplane/crossplane/blob/master/cluster/charts/crossplane-controllers/templates/package-manager-host-role.yaml). The package manager also will need cluster-scoped permissions, but that is included in the `ClusterRole` you defined below :+1:

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [inline design/design-doc-rbac-manager.md:605]: Just adding a note from out meeting that I would like for this to be discoverable on the revision itself, but it works either way

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [inline design/design-doc-rbac-manager.md:607]: I think we will name the SA the same name as the revision, so you may not need to list them all

**Review by hasheddan — COMMENTED** [review]
Thanks for this write-up @negz! I was thinking about how as long as the existing package manager is present that we will still be running it as privileged, but I think a simple solution would be to keep the existing package manager as its own `Deployment`, but make it optional, then build the new package manager into the existing Crossplane `Deployment`. Then we could eventually just remove the existing package manager altogether after a few cycles of deprecation. How does that sound?

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:140]: Yeah, this occurred to me this morning as well. I like your idea of keeping the existing package manager as a distinct deployment, rolling the new package manager into the core deployment, and granting the core deployment the access it needs (jobs, deployments, pods, etc).

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md]: What do you think about naming it `crossplane:provider:7f63e3661` (where `7f63e3661` is the revision name) to match the prefix of the RBAC roles the RBAC manager would generate for it?

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:139]: I've updated this role; let me know how it looks. I removed a few things like `deletecollection` and `list` that I suspect aren't really used, but I could be wrong.

**Comment by negz** [comment]
I think this design is ready for review - I've moved it out of draft.

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [reply design/design-doc-rbac-manager.md]: That sounds fine to me, although we can't include `:` as SA's must adhere to DNS-1123 subdomain naming. We could do `crossplane-provider-7f63e3661` or `crossplane.provider.7f63e3661` though.

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [inline design/design-doc-rbac-manager.md:415]:
```suggestion
# compositions. This allows them to discover and select an appropriate
```

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [inline design/design-doc-rbac-manager.md:639]: This would be deferred to a Crossplane user to build right?

**Review by hasheddan — COMMENTED** [review]
Added a few comments below, but the overall design LGTM

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md]: Ah, fair enough. This was just a minor optimisation so I'm happy to go with whatever makes the most sense to you. Using just the revision name seems fine to me. If we do go for a prefix I have a slight preference for hyphen separators.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:213]: Good point, they do. Will update.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:699]: Any preference on whether I name this just `7f63e3661` or `crossplane-provider-7f63e3661`? (Or something else?)

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:184]: Yeah I think this got lost when I moved some things around; originally this comment applied to both the the pod rules. Will fix.

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [reply design/design-doc-rbac-manager.md:699]: I am currently just naming it the hash, but it may be confusing if someone is looking at all of the SA's present in their cluster to just see a bunch of long hash names. I like `crossplane-provider-*`. If we adjust to a different name we can always circle back and update here :+1:

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:221]: My goal was to run providers with the least privileges they need to operate, and I don't think they need to be able to create or delete their managed resources or providers.

That said, I also tried to strike a balance between running with least privilege and writing super granular RBAC roles. For example if I was strictly going for least privilege I could remove all but RO access to provider configs, and remove Crossplane's ability to create RBAC roles.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:194]: Agreed. I'll break the `Role` out in this example to make things a little clearer. I might gloss over the differences between regular and hosted mode though - if I understand correctly the RBAC manager can be ignorant of those differences, which are handled by the Helm chart.

**Review by jbw976 — COMMENTED** [review]
This is a great write-up @negz! Should we also update https://github.com/crossplane/crossplane/blob/master/design/one-pager-packages-security-isolation.md to indicate that functionality is now obsolete (or at least out of date) and to now point to this document from there as the source of truth?

**Comment by negz** [comment]
> Should we also update https://github.com/crossplane/crossplane/blob/master/design/one-pager-packages-security-isolation.md to indicate that functionality is now obsolete (or at least out of date) and to now point to this document from there as the source of truth?

I don't think we have a super consistent approach to old design docs, but I agree that we should at least mark it `Status: Defunct`. I'd prefer to do that in a separate PR since we actually have quite a few docs that should be marked defunct at this stage. I've added https://github.com/crossplane/crossplane/issues/1711 to the 0.13 milestone to make sure we do that.

File: design/design-doc-rbac-manager.md
Comment (muvaf) [inline design/design-doc-rbac-manager.md:376]: Is the namespace-specific aggregation the only reason? Do you expect those roles to have cluster-scoped permissions at some point? (I'm not sure if we discussed this, probably did). If you don't expect cluster-scoped permissions to be added in the future, maybe RBAC manager could do the aggregation manually, i.e. look at annotations on the `Namespace` and update `Role`s in that namespace.

File: design/design-doc-rbac-manager.md
Comment (muvaf) [inline design/design-doc-rbac-manager.md:590]: Nit: maybe something with `offer` owrd, like `xrd-offered`?

File: design/design-doc-rbac-manager.md
Comment (muvaf) [inline design/design-doc-rbac-manager.md:734]:
```suggestion
* `--manage=all` - Include all functionality described in this design. This is the default.
```

`--manage=all` will be the default, right?

File: design/design-doc-rbac-manager.md
Comment (muvaf) [inline design/design-doc-rbac-manager.md:590]: Nit: maybe something with `offer` owrd, like `xrd-offered`?

**Review by muvaf — APPROVED** [review]
LGTM! I have a suggestion about doing the aggregation manually but I have a feeling that we already discussed it in one of many of our discussion, though I don't remember the arguments. So, if that's the case, maybe adding a few lines about it in the `Alternatives Considered` could be good.

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [reply design/design-doc-rbac-manager.md:194]: Yep that's correct :+1:

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:590]: I don't like using the term offered here - it conflates the act of creating a namespaced CRD with the act of managing RBAC roles that all that defined CR to be used in a namespace. "Offers a claim" is a platform builder concern, while "enables a claim in a namespace" is a platform operator concern.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:734]: Right.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:590]: I could also go for something like `rbac.crossplane.io/composites.example.org: claim-accepted`. I feel like this would fit pretty well, i.e.:

1. Platform builder offers a claim for this XRD.
1. Platform operator accepts the claim in this namespace.

Comment (negz) [reply design/design-doc-rbac-manager.md:590]: I think this comment is a dupe - resolving.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:639]: Sorry, I'm not quite sure what you're asking.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:376]: > Is the namespace-specific aggregation the only reason? Do you expect those roles to have cluster-scoped permissions at some point.

It is, and I don't.

> maybe RBAC manager could do the aggregation manually

I was on the fence about this too. On one hand I appreciate and share the desire to use a namespaced `Role` to represent a role that applies to a specific namespace. On the other hand I'm reluctant to 'reinvent' aggregation functionality that already exists. I might sketch out what it would look like if we used a `Role` after I address the other comments in this doc.

File: design/design-doc-rbac-manager.md
Comment (negz) [inline design/design-doc-rbac-manager.md:431]: On reflection I'm fairly sure this won't work. XRDs and Compositions are cluster scoped, and I believe rules pertaining to cluster scoped resources are ignored when a `ClusterRole` is bound to a subject in a particular namespace via a `RoleBinding`.

Comment (negz) [reply design/design-doc-rbac-manager.md]: @thephred I've introduced a new `crossplane-browse` cluster role to address this. I'm hopeful that this could be integrated into Upbound Cloud either by allowing granting 'browse' (along with admin, edit, or view) at the cluster scope or by automatically granting 'browse' at the cluster scope to anyone granted `edit` at the namespace scope.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md]: Take a look at the latest commit - I decided to go with having the RBAC manager manage namespaced roles by emulating/approximating an opinionated form of cluster role aggregation. I went back and forth a lot, but it seems like it's not really any more difficult to manually aggregate RBAC role rules (as opposed to manually managing aggregation rules), and it feels a little more 'idiomatic' than creating a cluster role for each namespace.

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [inline design/design-doc-rbac-manager.md:205]: I have ended up needing some additional permissions here (such as `watch`), but these could change depending on final details of package manager implementation, so I would defer to me updating this after that is merged.

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [inline design/design-doc-rbac-manager.md:190]: nitpick: also need to be able to create SA, but this can be updated later. I also considered if we might want the RBAC controller to create the SA for a `ProviderRevision`, then we could watch for its existence and know that when it existed the correct ClusterRole would already be bound and we would be safe to create the Deployment :thinking:

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [inline design/design-doc-rbac-manager.md:401]:
```suggestion
  name: crossplane-browse
```

File: design/design-doc-rbac-manager.md
Comment (hasheddan) [inline design/design-doc-rbac-manager.md:416]: I admittedly don't have as much knowledge of the `browse` role as the others, but I wonder if we would want a separate `ClusterRole` for each namespace such that a user can only "browse" the cluster-scoped types for which their is a claim type enabled for their namespace. Or maybe that is the intention here and I am misinterpreting? I would also be comfortable saying that this would be nice but the complexity can be deferred for the time-being.

**Review by hasheddan — APPROVED** [review]
One change needed, the rest are just comments that may or may not require changes. This doc will likely need some clean up after implementation is complete, so I am going ahead and giving approval :heavy_check_mark: Thanks for describing this complex controller in a digestible manner @negz!

File: design/design-doc-rbac-manager.md
Comment (muvaf) [reply design/design-doc-rbac-manager.md]: Nice! I love that we're able to do that. There will be some complexity added to the controller code but I believe that's manageable given the cost of the alternative. Thanks!

**Comment by thephred** [comment]
LGTM

I believe that a user that has been given rights to view/edit a CompositeResource claim should be able to see the CompositeResource that is created on its behalf as well as list/view CompositeResourceDefinitions.

File: design/design-doc-rbac-manager.md
Comment (lukeweber) [inline design/design-doc-rbac-manager.md:636]: Suggestion: If one happens to uninstall or upgrade packages these annotations might become out of date. As well, if there's a typo, it wouldn't be obvious that it might fail to provision permissions for a type, or that it was an out of date permission grant.

For someone building a UI around this like Upbound cloud, listing a particular Kind, and seeing the status of whether the permission could be applied, or if it's deprecated/no longer existing would be a lot easier to work with, than a totally different component reading the namespace annotations, parsing the xrd out of them, and comparing these names with what's installed on the server to know what's still valid.

**Review by lukeweber — APPROVED** [review]
Generally looks good to me. One small comment that I think a separate type is probably more useful for Upbound for building a UI around these permissions as the RBAC controller could give useful feedback, but we can make it work either way.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:636]: @lukeweber Could you sketch me some YAML representing the resource you're thinking of?

Comment (lukeweber) [reply design/design-doc-rbac-manager.md:636]: The only point was just that the calculation of whether you were able to find such an xrd will technically already be known by the RBAC controller and might make for a better DX, but as you noted it's another type. I'll defer to your judgement here. Just a suggestion :)

```yaml
apiVersion: rbac.authorization.crossplane.io/v1
kind: XRDGrant
metadata:
  name: composites.example.org
  namespace: foo
spec:
  xrd: composites.example.org
status:
  phases: ready , failed
  conditions: xrdMissing, createdRBAC, noPermissions
```

Comment (negz) [reply design/design-doc-rbac-manager.md:636]: Got it. It seems like the key value here is that we could use the new type's status to reflect whether the intent to accept/grant a claim in a namespace was successful. I'd like to stick with the current design for now, but we can introduce this type if annotations prove to be a poor UX. I believe it's fairly common in Kubernetes for something to start as an annotation and later grow into a real field/type.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:416]: Yeah, let's avoid that complexity for now. I'd like to avoid generating "namespace aligned ClusterRoles" if possible. I'm hopeful that this broad 'browse' role (or even the view role) will be appropriate for 80% of folks - keep in mind that there's always the fallback option of authoring and granting your own roles.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:205]: I'll add list and watch for now. I agree we can update this later to more accurately reflect reality as we understand it better.

File: design/design-doc-rbac-manager.md
Comment (negz) [reply design/design-doc-rbac-manager.md:190]: I've updated these rules to add create SA.

Regarding the RBAC manager creating SAs, I think the separation of concerns is perhaps a little cleaner if the package manager owns that task. That way you can still install a and run provider if you're not running the RBAC manager - you just have to make sure some other entity grants RBAC access (or that RBAC is disabled).

**Pull request merged**
